File: delta/data/frontend/zcr.py
# Copyright (C) 2017 Beijing Didi Infinity Technology and Development Co.,Ltd.
# All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""This model extracts zcr features per frame."""

import delta.compat as tf

from core.ops import py_x_ops
from delta.utils.hparam import HParams
from delta.data.frontend.base_frontend import BaseFrontend


class Zcr(BaseFrontend):
  """
  Compute ZCR features respectively，and concate them. Return
  a tensor with shape (1, num_frames).
  """

  def __init__(self, config: dict):
    super().__init__(config)

  @classmethod
  def params(cls, config=None):
    """
    Set params.
    :param config:contains three optional parameters:
        --sample_rate       : Waveform data sample frequency (must match the waveform
                             file, if specified there). (float, default = 16000)
        --window_length		 : Window length in seconds. (float, default = 0.025)
        --frame_length		 : Hop length in seconds. (float, default = 0.010)
    :return: An object of class HParams, which is a set of hyperparameters as
            name-value pairs.
    """

    window_length = 0.025
    frame_length = 0.010
    sample_rate = 16000

    hparams = HParams(cls=cls)
    hparams.add_hparam('window_length', window_length)
    hparams.add_hparam('frame_length', frame_length)
    hparams.add_hparam('sample_rate', sample_rate)

    if config is not None:
      hparams.override_from_dict(config)

    return hparams

  def call(self, audio_data, sample_rate=None):
    """
    Calculate the zero-crossing rate of speech.
    :param audio_data: the audio signal from which to compute spectrum.
                      Should be an (1, N) tensor.
    :param sample_rate: [option]the samplerate of the signal we working with,
                        default is 16kHz.
    :return: A tensor with shape (1, num_frames), containing zero-crossing rate of
            every frame in speech.
    """

    p = self.config
    with tf.name_scope('zcr'):

      if sample_rate == None:
        sample_rate = tf.constant(p.sample_rate, dtype=tf.int32)

      assert_op = tf.assert_equal(
          tf.constant(p.sample_rate), tf.cast(sample_rate, dtype=tf.int32))
      with tf.control_dependencies([assert_op]):

        sample_rate = tf.cast(sample_rate, dtype=float)
        zcr = py_x_ops.zcr(
            audio_data,
            sample_rate,
            window_length=p.window_length,
            frame_length=p.frame_length)

        return zcr

File: transform/kernels/zcr.cc
/* Copyright (C) 2017 Beijing Didi Infinity Technology and Development Co.,Ltd.
All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "support_functions.h"
#include "zcr.h"

#define ABS(a) (a > 0 ? a : -a)

namespace delta {
const float window_length_sec = 0.025;
const float frame_length_sec = 0.010;

ZCR::ZCR() {
  window_length_sec_ = window_length_sec;
  frame_length_sec_ = frame_length_sec;
  pf_ZCR = NULL;
}

ZCR::~ZCR() { free(pf_ZCR); }

void ZCR::set_window_length_sec(float window_length_sec) {
  window_length_sec_ = window_length_sec;
}

void ZCR::set_frame_length_sec(float frame_length_sec) {
  frame_length_sec_ = frame_length_sec;
}

int ZCR::init_zcr(int input_size, float sample_rate) {
  f_SamRat = sample_rate;
  i_WinLen = static_cast<int>(window_length_sec_ * f_SamRat);
  i_FrmLen = static_cast<int>(frame_length_sec_ * f_SamRat);
  i_NumFrm = (input_size - i_WinLen) / i_FrmLen + 1;

  pf_ZCR = static_cast<float*>(malloc(sizeof(float) * i_NumFrm));

  return 1;
}

int ZCR::proc_zcr(const float* mic_buf) {
  int n, k;
  float* win = static_cast<float*>(malloc(sizeof(float) * i_WinLen));
  float tmp;

  for (n = 0; n < i_NumFrm; n++) {
    for (k = 0; k < i_WinLen; k++) {
      win[k] = mic_buf[n * i_FrmLen + k];
    }
    pf_ZCR[n] = 0.0;
    for (k = 1; k < i_WinLen; k++) {
      tmp = static_cast<float>(compute_sign(win[k]) - compute_sign(win[k - 1]));
      pf_ZCR[n] = pf_ZCR[n] + ABS(tmp) / 2 / i_WinLen;
    }
  }

  free(win);
  return 1;
}

int ZCR::get_zcr(float* output) {
  memcpy(output, pf_ZCR, sizeof(float) * i_NumFrm);

  return 1;
}

int ZCR::write_zcr() {
  FILE* fp;
  fp = fopen("zero_cross_rate.txt", "w");
  int n;
  for (n = 0; n < i_NumFrm; n++) {
    fprintf(fp, "%4.6f\n", pf_ZCR[n]);
  }
  fclose(fp);
  return 1;
}
}  // namespace delta
